mailbox/opensearch/src/main/java/org/apache/james/mailbox/opensearch/events/OpenSearchListeningMessageSearchIndex.java

```diff
@@ -131,7 +157,126 @@ public EnumSet<SearchCapabilities> getSupportedCapabilities(EnumSet<MessageCapab
             SearchCapabilities.AttachmentFileName,
             SearchCapabilities.PartialEmailMatch);
     }
-    
+
+    @Override
+    public Mono<Void> reactiveEvent(Event event) {
+        MailboxSession systemSession = sessionProvider.createSystemSession(event.getUsername());
+        return handleMailboxEvent(event, systemSession, (MailboxEvents.MailboxEvent) event)
+            .then(Mono.fromRunnable(() -> factory.endProcessingRequest(systemSession)));
+    }
+
+    private Mono<Void> handleMailboxEvent(Event event, MailboxSession session, MailboxEvents.MailboxEvent mailboxEvent) {
+        MailboxId mailboxId = mailboxEvent.getMailboxId();
+
+        if (event instanceof MailboxEvents.Added) {
+            return handleAddedEvent(session, (MailboxEvents.Added) event, mailboxId);
+        } else if (event instanceof MailboxEvents.Expunged) {
+            return handleExpungedEvent((MailboxEvents.Expunged) event, session, mailboxId);
+        } else if (event instanceof MailboxEvents.FlagsUpdated) {
+            MailboxEvents.FlagsUpdated flagsUpdated = (MailboxEvents.FlagsUpdated) event;
+            return update(session, mailboxId, flagsUpdated.getUpdatedFlags());
+        } else if (event instanceof MailboxEvents.MailboxDeletion) {
+            return deleteAll(session, mailboxId);
+        } else {
+            return Mono.empty();
+        }
+    }
+
+    private Mono<Void> handleExpungedEvent(MailboxEvents.Expunged expunged, MailboxSession session, MailboxId mailboxId) {
+        if (messageIndexNewBehavior) {
```

Comment:
I hate that if.

Dirty code.


Better idea:

```
interface IndexingStrategy {
    Mono<Void> handleExpungedEvent(MailboxEvents.Expunged expunged, MailboxSession session, MailboxId mailboxId);
     Mono<Void> handleAddedEvent(MailboxSession session, MailboxEvents.Added addedEvent,  MailboxId mailboxId)
}

class NaiveIndexingStrategy { ... }

class OptimizedIndexingStrategy { ... }
```

 -> This avoids leaking the fact that there is two behaviour everywhere
 -> Makes changing the indexing strategy easier (could those even be injected in the future ???)
 -> And it would be SUPER EASY to add a new one in the future
 -> And we avoid 5 indent depth in some method (wooohooo)

Reply:
it really better, thanks